### ISSUE TITLE
VimwikiTOC is broken against headers with link

### ISSUE THREAD

**Issue by ryujinno** [vimwiki/vimwiki#182]
VimwikiTOC is broken against headers with link.
This wiki reproduces the bug.

```
= Contents =
  - [[#[[http://example.com/|Example Domain]]|[[http://example.com/|Example Domain]]]]

= [[http://example.com/|Example Domain]] =
```

This bug is not reproduced with old %toc.

**Comment by Nudin** [comment]
A `#` in a title also breaks VimwikiTOC

**Comment by ranebrown** [comment]
Headers with parenthesis also don't work for markdown syntax but I would guess that will be fixed with #619

**Comment by riazrizvi** [comment]
It looks like `VimwikiTOC`  _eats_ lines containing any `rxBulletList` symbol, until it finds a non-empty line that doesn't. So if the first heading contains for example a `#` anywhere, then it is still considered *under* `=Contents=` and it will be deleted. If I understand the function of `VimwikiTOC`, it should stop replacing when a non-empty line does not have a first non-blank character = either `#`, `-`, `*`. 

The problem can be fixed by replacing the code to generate `rxBulletList` in `autoload/vimwiki/vars.vim:vimwiki#vars#populate_syntax_vars` with this:
```
  if !empty(g:vimwiki_syntax_variables[a:syntax].bullet_types)
    let g:vimwiki_syntax_variables[a:syntax].rxListBullet =
          \ join( map( map(copy(g:vimwiki_syntax_variables[a:syntax].bullet_types),
          \'vimwiki#u#escape(v:val).'
          \ .'repeat("\\+", g:vimwiki_syntax_variables[a:syntax].recurring_bullets)'
          \ ), 'escape("^s*","s") . (v:val)') , '\|')
  else
    "regex that matches nothing
    let g:vimwiki_syntax_variables[a:syntax].rxListBullet = '$^'
  endif

```
It prepends a `^\s*` to every `rxBulletList` regex entry, which is used in `vimwiki#base#update_listing_in_buffer` to replace old TOC.

Since `VimwikiGenerateLinks` also uses the same regex via `vimwiki#base#update_listing_in_buffer`, I think it has a similar problem and this will also address that.

The nested `map` function is a bit complicated IMO. Might be easier to understand and debug if the whole thing was a transparent loop?

**Comment by tinmarino** [comment]
Fixed: 69ead3b  (see commit message for more info)
